modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -0,0 +1,331 @@
+
+        assert self.context, "EvaluationExecutor not initialized"
+
+        df_triggers = self.context.tracking_dfs.get(PipelineStage.HANDLE_SINGLE_TRIGGER.name)
+        df_store_models = self.context.tracking_dfs.get(PipelineStage.STORE_TRAINED_MODEL.name)
+        if df_triggers is None or df_store_models is None:
+            logger.warning("Could not run evaluations as no training was found in the tracking info dataframes")
+            return SupervisorLogs(stage_runs=[])
+        df_trainings = df_triggers[["trigger_id", "first_timestamp", "last_timestamp"]].merge(
+            df_store_models[["id_model", "training_id", "trigger_id"]], on="trigger_id"
+        )
+
+        eval_requests: list[EvalRequest] = []
+        for eval_handler in self.eval_handlers:
+            if eval_handler.config.execution_time != "after_pipeline":
+                continue
+
+            handler_eval_requests = eval_handler.get_eval_requests_after_pipeline(df_trainings=df_trainings)
+            eval_requests += handler_eval_requests
+
+        num_workers = None
+        if self.config.supervisor:
+            num_workers = self.config.supervisor.after_pipeline_evaluation_workers
+
+        logs = self._launch_evaluations_async(
+            eval_requests,
+            parent_log=StageLog(
+                id=PipelineStage.EVALUATE_SINGLE.name,
+                start=datetime.datetime.now(),
+                batch_idx=-1,
+                sample_idx=-1,
+                sample_time=-1,
+                trigger_idx=-1,
+            ),
+            eval_status_queue=eval_status_queue,
+            num_workers=num_workers,
+        )
+        return logs
+
+    # -------------------------------------------------------------------------------- #
+    #                                     Internal                                     #
+    # -------------------------------------------------------------------------------- #
+
+    def _launch_evaluations_async(
+        self,
+        eval_requests: list[EvalRequest],
+        parent_log: StageLog,
+        eval_status_queue: Queue,
+        num_workers: int | None = None,
+    ) -> SupervisorLogs:
+        """Creates a thread pool to launch evaluations asynchronously.
+
+        Args:
+            eval_requests: The evaluation requests to launch.
+            num_workers: The number of workers to use.
+        """
+
+        tasks: list[Future[StageLog]] = []
+        logs = SupervisorLogs()
+
+        def worker_func(eval_req: EvalRequest) -> StageLog:
+            single_log = StageLog(
+                id=PipelineStage.EVALUATE_SINGLE.name,
+                start=datetime.datetime.now(),
+                batch_idx=parent_log.batch_idx,
+                sample_idx=parent_log.sample_idx,
+                sample_time=parent_log.sample_time,
+                trigger_idx=parent_log.trigger_idx,
+            )
+            epoch_nanos_start = current_time_micros()
```

Comment:
nanos or micros :D?

Reply:
Ah thanks, that was already wrong in the pipeline_executor xD